airbyte-commons/src/main/java/io/airbyte/commons/util/ParallelCompositeIterator.java

```diff
@@ -0,0 +1,276 @@
+  private static final Integer MAX_THREADS = 10;
+  private static final Integer MIN_THREADS = 1;
+
+  private final Optional<Consumer<AirbyteStreamStatusHolder>> airbyteStreamStatusConsumer;
+  private final ExecutorService executorService;
+  private boolean hasClosed;
+  private final BlockingQueue<InternalMessage<T>> internalQueue;
+  private final Set<AutoCloseableIterator<T>> iterators;
+  private boolean started = false;
+  private int closedCount = 0;
+
+  ParallelCompositeIterator(final Collection<AutoCloseableIterator<T>> iterators,
+                            final Consumer<AirbyteStreamStatusHolder> airbyteStreamStatusConsumer) {
+    this.airbyteStreamStatusConsumer = Optional.ofNullable(airbyteStreamStatusConsumer);
+    this.iterators = iterators != null ? new HashSet<>(iterators) : new HashSet<>();
+    /*
+     * TODO What is the correct upper bound here? It should probably be provided by the code that
+     * creates this iterator to account for the different types of connectors and their limitations. For
+     * instance, a JDBC-based source connector should probably set the "MAX_THREADS" to the maximum
+     * number of connections in the database connection pool used by the source. For API-based source
+     * connectors, this should be the maximum number of concurrent connections from one host/IP address
+     * that the API supports.
+     */
+    executorService = Executors.newFixedThreadPool(computeThreadPoolSize());
+    /*
+     * TODO Expose the max capacity as an advanced connector setting. This would allow the end user to
+     * configure the trade-off between throughput and resource usage (memory to handle the messages on
+     * the queue). Records may have arbitrary sizes requiring this value to be tweaked based on the
+     * shape of the data being read.
+     */
+    internalQueue = new LinkedBlockingQueue<>(MAX_CAPACITY);
+    this.hasClosed = false;
+  }
+
+  /**
+   * Starts the parallel processing of the underlying iterators by submitting each iterator to the
+   * internal thread pool. If the iterator has already started parallel processing, the call to this
+   * method will be ignored.
+   */
+  public void start() {
+    if (!started) {
+      LOGGER.debug("Starting parallel consumption of streams...");
+      iterators.forEach(iterator -> {
+        executorService.submit(() -> consumeFromStream(iterator));
+      });
+      started = true;
+      LOGGER.debug("Parallel stream consumption started.");
+    } else {
+      LOGGER.debug("Parallel stream consumption already started.");
+    }
+  }
+
+  @Override
+  protected T computeNext() {
+    try {
+      assertHasNotClosed();
+
+      if (!started) {
+        start();
+      }
+
+      if (iterators.isEmpty()) {
+        return endOfData();
+      }
+
+      // Block while waiting for the next result to be available
+      final T result = internalQueue.take().data();
+      LOGGER.debug("Fetched next result {} from internal queue.", result);
```

Comment:
NIT: this should not make it to prod, it activates the possibility of having client data in the logs.